### ISSUE TITLE
Add calculation for Terrestrial Coupling Index (TCI)

### ISSUE THREAD

**Issue by DanielAdriaansen** [dtcenter/METcalcpy#364]
## Describe the New Feature ##
I am updating the METplus Wrappers use case here: https://metplus.readthedocs.io/en/develop/generated/model_applications/land_surface/PointStat_fcstCESM_obsFLUXNET2015_TCI.html#sphx-glr-generated-model-applications-land-surface-pointstat-fcstcesm-obsfluxnet2015-tci-py. That use case does Python embedding for both the forecast and observation data. For the observation data, it was previously using pre-computed TCI data but now it will calculate TCI from raw observation data. The TCI calculation should be shared for the forecast and observation data since it is mathematically the same.

### Acceptance Testing ###
*List input data types and sources.*
*Describe tests required for new functionality.*

### Time Estimate ###
*Estimate the amount of work required here.*
*Issues should represent approximately 1 to 3 days of work.*

### Sub-Issues ###
Consider breaking the new feature down into sub-issues.
- [ ] *Add a checkbox for each sub-issue here.*

### Relevant Deadlines ###
*List relevant project deadlines here or state NONE.*

### Funding Source ###
*Define the source of funding and account keys here or state NONE.*

### Assignee ###
- [ ] Select **engineer(s)** or **no engineer** required
- [ ] Select **scientist(s)** or **no scientist** required

### Labels ###
- [ ] Select **component(s)**
- [ ] Select **priority**
- [ ] Select **requestor(s)**

### Projects and Milestone ###
- [ ] Select **Repository** and/or **Organization** level **Project(s)** or add **alert: NEED CYCLE ASSIGNMENT** label
- [ ] Select **Milestone** as the next official version or **Future Versions**

## Define Related Issue(s) ##
Consider the impact to the other METplus components.
- [ ] [METplus](https://github.com/dtcenter/METplus/issues/new/choose), [MET](https://github.com/dtcenter/MET/issues/new/choose), [METdataio](https://github.com/dtcenter/METdataio/issues/new/choose), [METviewer](https://github.com/dtcenter/METviewer/issues/new/choose), [METexpress](https://github.com/dtcenter/METexpress/issues/new/choose), [METcalcpy](https://github.com/dtcenter/METcalcpy/issues/new/choose), [METplotpy](https://github.com/dtcenter/METplotpy/issues/new/choose)

## New Feature Checklist ##
See the [METplus Workflow](https://metplus.readthedocs.io/en/latest/Contributors_Guide/github_workflow.html) for details.
- [ ] Complete the issue definition above, including the **Time Estimate** and **Funding source**.
- [ ] Fork this repository or create a branch of **develop**.
Branch name: `feature_<Issue Number>_<Description>`
- [ ] Complete the development and test your changes.
- [ ] Add/update log messages for easier debugging.
- [ ] Add/update unit tests.
- [ ] Add/update documentation.
- [ ] Add any new Python packages to the [METplus Components Python Requirements](https://metplus.readthedocs.io/en/develop/Users_Guide/overview.html#metplus-components-python-requirements) table.
- [ ] Push local changes to GitHub.
- [ ] Submit a pull request to merge into **develop**.
Pull request: `feature <Issue Number> <Description>`
- [ ] Define the pull request metadata, as permissions allow.
Select: **Reviewer(s)** and **Development** issues
Select: **Repository** level development cycle **Project** for the next official release
Select: **Milestone** as the next official version
- [ ] Iterate until the reviewer(s) accept and merge your changes.
- [ ] Delete your fork or branch.
- [ ] Close this issue.

**Comment by DanielAdriaansen** [comment]
I am not sure if this is "contributed" or a "diagnostic":
https://metcalcpy.readthedocs.io/en/latest/Contributors_Guide/index.html#organization-of-code-in-the-github-repository

Thoughts?

**Comment by DanielAdriaansen** [comment]
> I am not sure if this is "contributed" or a "diagnostic": https://metcalcpy.readthedocs.io/en/latest/Contributors_Guide/index.html#organization-of-code-in-the-github-repository
> 
> Thoughts?

I think I could put this under `contributed/land_sfc` to be consistent with the METplus use case `model_applications` category: https://metplus.readthedocs.io/en/latest/generated/model_applications/index.html#land-surface?

Or, I could see a `diagnostics/land_sfc` directory, but I see the `diagnostics` directory is empty so maybe this is not the way to go yet.

**Comment by bikegeek** [comment]
If it truly is a diagnostic, then it should go under the diagnostics
directory.  We put that directory as a placeholder for future diagnostic
code.  Since you are the scientist on the issue, I'll defer to your
decision.  The contributed directory was created for outside collaborators,
whose code didn't necessarily conform with our coding structure.

**Comment by DanielAdriaansen** [comment]
> The contributed directory was created for outside collaborators, whose code didn't necessarily conform with our coding structure.

OK this helps me distinguish it a lot! The code I am adding is very small (30 lines?) and I have the freedom to write it in whatever standards we choose so I think ultimately this will land in `diagnostics/land_surface`. I am double checking with some other folks on the project. Thank you @bikegeek!

**Comment by bikegeek** [comment]
That's great, this code will be a great addition to METcalcpy and we
finally have something that is available in the diagnostics 'library'.

**Comment by DanielAdriaansen** [comment]
@bikegeek one question I have is about structure under `diagnostics`. One option is to have a `land_surface.py` file, the other is to have `diagnostics/land_surface/land_surface_diagnostic_name.py` structure. Can you advise?

**Comment by bikegeek** [comment]
I like the diagnostics/land_surface/land_surface_diagnostic_name.py
structure, it allows the addition of future diagnostics that aren't
land_surface in nature.

**Comment by DanielAdriaansen** [comment]
OK, then if I had this structure:
`diagnostics/land_surface/terrestrial_coupling_index.py`

and inside `terrestrial_coupling_index.py` there is a function called `calc_tci()`, can you help me formulate the import statement in my Python script?

Would it be:
```
from metcalcpy.diagnostics import land_surface as land_sfc
tci = land_sfc.calc_tci()
```

I don't think that's right, I think that's if it was `diagnostics/land_surface.py`. The extra directory confuses me for some reason but I am probably just forgetting how it works!

**Comment by bikegeek** [comment]
If the terrestrial_coupling_index a class:

class TerrestrialCouplingIndex():

   def  __init__():
       # stuff for initializing the TerrestrialCouplingIndex object

  def calc_tci():
      # stuff goes here


-----
then the import would look like:

from metcalcpy.diagnostics.land_surface.terrestrial_couplint_index import
TerrestrialCouplingIndex as tci

and accessing the methods from the terrestrial_coupling_index.py module:

    # Create an instance of the TerrestrialCouplingIndex object with
specific initialization params (like input filename, whatever)
    tci_obj = tci.TerrestrialCouplingIndex(param1, param2)

    # access the methods via the tci_obj object/instance
    tc_index  = tci_obj.calc_ti()


I'm bad at naming variables, I hope I haven't confused you with the poor
choice of var names.

**Comment by anewman89** [comment]
@bikegeek @DanielAdriaansen This issue should use CLASP - 7790111 for charging.

**Comment by DanielAdriaansen** [comment]
Closed via #369
